File: src/vpc/runtime/premiumDeckUrl.js
"use strict";

const { S3Client, GetObjectCommand } = require("@aws-sdk/client-s3");
const { getSignedUrl } = require("@aws-sdk/s3-request-presigner");
const { logInfo, logError } = require("../../common/log");
const { pool } = require("../db/pg");

const PREMIUM_URL_IMPL = "premiumDeckUrl-v7"; // ✅ bump so you can verify deploy

function safeSlug(s) {
  const v = String(s || "").trim();
  if (!/^[a-z0-9-]+$/i.test(v)) return null;
  return v.toLowerCase();
}

function toInt(v, def) {
  const n = Number(v);
  return Number.isFinite(n) && n > 0 ? Math.floor(n) : def;
}

function toBool(v) {
  const s = String(v ?? "").trim().toLowerCase();
  return s === "1" || s === "true" || s === "yes";
}

function pickHeader(event, name) {
  const h = event?.headers || {};
  return String(h[name] || h[name.toLowerCase()] || "").trim();
}

// Some API gateways / proxies rewrite header names; be defensive.
function readDevBypassHeader(event) {
  const candidates = ["x-dev-bypass", "X-Dev-Bypass", "x_dev_bypass", "X_DEV_BYPASS"];
  for (const k of candidates) {
    const v = pickHeader(event, k);
    if (v) return v;
  }
  return "";
}

function parseAllowlist(raw) {
  const s = String(raw || "").trim();
  if (!s) return new Set();
  return new Set(
    s
      .split(",")
      .map((x) => x.trim())
      .filter(Boolean),
  );
}

function withAbort(ms) {
  const ac = new AbortController();
  const t = setTimeout(() => ac.abort(new Error(`abort_timeout_${ms}ms`)), ms);
  return { signal: ac.signal, cancel: () => clearTimeout(t) };
}

async function isPremiumFromDb(userSub) {
  const p = pool();
  if (!p) {
    return {
      ok: false,
      premium: false,
      env: null,
      productId: null,
      expiresAtMs: null,
      reason: "NO_DB_POOL",
    };
  }

  const r = await p.query(
    `
    select premium_active, premium_env, product_id, expires_at_ms
    from user_premium_state
    where app_user_id = $1
    limit 1;
    `,
    [String(userSub)],
  );

  const row = r.rows?.[0] || null;
  const expiresAtMs = row?.expires_at_ms != null ? Number(row.expires_at_ms) : null;
  const nowMs = Date.now();

  // prefer expires_at_ms if present
  const premium = expiresAtMs != null ? expiresAtMs > nowMs : !!row?.premium_active;

  const env = row?.premium_env || null; // sandbox/production/none
  const productId = row?.product_id || null;

  return {
    ok: true,
    premium,
    env,
    productId,
    expiresAtMs,
    reason: premium ? "ACTIVE" : "NOT_ACTIVE",
  };
}

async function handlePremiumDeckUrl({ event, method, query, res, auth }) {
  if (method !== "GET") return res.methodNotAllowed();

  const traceId = res.traceId || null;

  const slug = safeSlug(query?.slug);
  if (!slug) return res.badRequest("Missing/invalid slug");

  const apiEnv = String(process.env.API_ENV || "").trim().toLowerCase();
  const isProdEnv = apiEnv === "production";

  // ✅ always require login for this endpoint (even dev bypass)
  const userSub = String(auth?.userSub || "").trim() || null;
  if (!userSub) return res.forbidden("Requires login");

  // ===== DEV BYPASS (server-side) =====
  // enable when ALL:
  // - API_ENV=development (hard safety)
  // - ALLOW_DEV_PREMIUM=1
  // - client sends dev=1 OR x-dev-bypass=1 (dev route will auto-inject)
  // - userSub is in DEV_BYPASS_ALLOWLIST_SUBS (hard safety)
  const devFlag = String(query?.dev || "") === "1";
  const devHeaderRaw = readDevBypassHeader(event);
  const devHeader = devHeaderRaw === "1";

  const allowlist = parseAllowlist(process.env.DEV_BYPASS_ALLOWLIST_SUBS);
  const allowlistEnabled = allowlist.size > 0;
  const allowlisted = allowlistEnabled ? allowlist.has(userSub) : false;

  const allowDevBypass =
    !isProdEnv &&
    String(process.env.ALLOW_DEV_PREMIUM || "") === "1" &&
    (devFlag || devHeader) &&
    allowlisted;

  // ✅ Always log bypass evaluation (so you can see why it didn't work)
  logInfo(
    JSON.stringify({
      traceId,
      impl: PREMIUM_URL_IMPL,
      step: "dev_bypass_eval",
      slug,
      apiEnv: apiEnv || null,
      devFlag,
      devHeader,
      devHeaderRaw: devHeaderRaw || null,
      allowDevBypass,
      allowlistEnabled,
      allowlisted,
      ALLOW_DEV_PREMIUM: process.env.ALLOW_DEV_PREMIUM || null,
      DISALLOW_SANDBOX_PREMIUM: process.env.DISALLOW_SANDBOX_PREMIUM || null,
    }),
  );

  // Config
  const region = process.env.AWS_REGION || "ap-southeast-2";
  const contentBucket = process.env.CONTENT_BUCKET;
  const premiumBucket = process.env.PREMIUM_BUCKET;

  const contentPrefix = (process.env.CONTENT_PREFIX || "content").replace(/^\/+|\/+$/g, "");
  const premiumPrefix = (process.env.PREMIUM_PREFIX || "premium").replace(/^\/+|\/+$/g, "");
  const expiresInSec = toInt(process.env.PREMIUM_URL_EXPIRES_SEC, 60);

  const s3GetTimeoutMs = toInt(process.env.S3_GET_TIMEOUT_MS, 6000);
  const s3BodyTimeoutMs = toInt(process.env.S3_BODY_TIMEOUT_MS, 6000);

  if (!contentBucket) return res.error400("Missing env CONTENT_BUCKET");
  if (!premiumBucket) return res.error400("Missing env PREMIUM_BUCKET");

  // 0) premium check (skip only when allowDevBypass)
  let premiumInfo = null;
  if (!allowDevBypass) {
    try {
      logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "premium_check_start", userSub }));

      premiumInfo = await isPremiumFromDb(userSub);

      logInfo(
        JSON.stringify({
          traceId,
          impl: PREMIUM_URL_IMPL,
          step: "premium_check_done",
          premium: !!premiumInfo?.premium,
          premiumEnv: premiumInfo?.env || null,
          expiresAtMs: premiumInfo?.expiresAtMs ?? null,
        }),
      );

      if (!premiumInfo.premium) return res.forbidden("Requires premium entitlement");

      // ✅ enforce sandbox ban ONLY when not dev bypass
      if (toBool(process.env.DISALLOW_SANDBOX_PREMIUM)) {
        if (String(premiumInfo.env || "").toLowerCase() === "sandbox") {
          return res.forbidden("Sandbox premium not allowed in this environment");
        }
      }
    } catch (e) {
      logError("premium check failed:", e);
      return res.error500(e);
    }
  } else {
    logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "dev_bypass_enabled", slug, userSub }));
  }

  const s3 = new S3Client({ region });

  try {
    // 1) load manifest
    const manifestKey = `${contentPrefix}/manifest.json`;

    logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "s3_get_manifest_start", manifestKey }));

    const a1 = withAbort(s3GetTimeoutMs);
    let obj;
    try {
      obj = await s3.send(new GetObjectCommand({ Bucket: contentBucket, Key: manifestKey }), { abortSignal: a1.signal });
    } finally {
      a1.cancel();
    }

    logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "s3_get_manifest_done" }));

    // 2) read body safely
    logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "manifest_read_start" }));

    const a2 = withAbort(s3BodyTimeoutMs);
    let body;
    try {
      if (!obj?.Body || typeof obj.Body.transformToString !== "function") {
        throw new Error("S3BodyMissingTransformToString");
      }
      body = await obj.Body.transformToString("utf-8", { abortSignal: a2.signal });
    } finally {
      a2.cancel();
    }

    logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "manifest_read_done", bytes: body?.length || 0 }));

    const manifest = JSON.parse(body);

    // 3) find deck
    const decks = Array.isArray(manifest?.decks) ? manifest.decks : [];
    const deck = decks.find((d) => String(d?.slug || "").toLowerCase() === slug);
    if (!deck) return res.notFound(`Deck not found in manifest: ${slug}`);

    const downloadMode = String(deck.downloadMode || "").toLowerCase();
    const tier = String(deck.tier || "").toLowerCase();

    if (!(downloadMode === "auth" || tier === "premium")) {
      return res.badRequest(`Not a premium deck: ${slug}`);
    }

    const buildId = String(deck.buildId ?? deck.version ?? "").trim();
    if (!buildId || buildId === "coming") return res.badRequest(`Deck not publishable: ${slug}`);

    // 4) presign full deck from PREMIUM bucket
    const premiumKey = `${premiumPrefix}/decks/${slug}/builds/${buildId}/deck.json`;

    logInfo(
      JSON.stringify({
        traceId,
        impl: PREMIUM_URL_IMPL,
        step: "presign_start",
        slug,
        buildId,
        premiumKey,
        allowDevBypass,
      }),
    );

    const url = await getSignedUrl(
      s3,
      new GetObjectCommand({
        Bucket: premiumBucket,
        Key: premiumKey,
        ResponseContentType: "application/json",
        ResponseCacheControl: "no-cache",
      }),
      { expiresIn: expiresInSec },
    );

    logInfo(JSON.stringify({ traceId, impl: PREMIUM_URL_IMPL, step: "presign_done" }));

    return res.ok({
      slug,
      buildId,
      expiresInSec,
      url,
      devBypass: allowDevBypass,
    });
  } catch (e) {
    logError("handlePremiumDeckUrl error:", e);
    return res.error500(e);
  }
}

module.exports = { handlePremiumDeckUrl };

File: src_C/Vpc/Runtime/PremiumDeckUrl.cs
using System.Globalization;
using System.Text.Json;
using Amazon;
using Amazon.Lambda.APIGatewayEvents;
using Amazon.S3;
using Amazon.S3.Model;
using RecallSmith.Lambda.Common;
using RecallSmith.Lambda.Db;

namespace RecallSmith.Lambda.Vpc.Runtime;

public static class PremiumDeckUrl
{
  private const string PremiumUrlImpl = "premiumDeckUrl-v7";

  private static string GetRegion()
  {
    var r = (Environment.GetEnvironmentVariable("AWS_REGION") ?? string.Empty).Trim();
    return string.IsNullOrEmpty(r) ? "ap-southeast-2" : r;
  }

  private static Lazy<AmazonS3Client> S3Lazy =
    new(() => new AmazonS3Client(RegionEndpoint.GetBySystemName(GetRegion())));

  // Used by SnapStart runtime hooks to ensure we don't reuse pre-snapshot network state.
  public static void Reset()
  {
    if (S3Lazy.IsValueCreated)
    {
      try { S3Lazy.Value.Dispose(); } catch { /* best-effort */ }
    }
    S3Lazy = new Lazy<AmazonS3Client>(() => new AmazonS3Client(RegionEndpoint.GetBySystemName(GetRegion())));
  }

  private static string? SafeSlug(string? s)
  {
    var v = (s ?? string.Empty).Trim();
    if (!System.Text.RegularExpressions.Regex.IsMatch(v, "^[a-z0-9-]+$", System.Text.RegularExpressions.RegexOptions.IgnoreCase))
    {
      return null;
    }
    return v.ToLowerInvariant();
  }

  private static int ToInt(string? v, int def)
  {
    if (int.TryParse(v, NumberStyles.Integer, CultureInfo.InvariantCulture, out var n) && n > 0) return n;
    return def;
  }

  private static bool ToBool(string? v)
  {
    var s = (v ?? string.Empty).Trim().ToLowerInvariant();
    return s is "1" or "true" or "yes";
  }

  private static string PickHeader(LambdaRequest req, string name)
  {
    return req.Headers.TryGetValue(name, out var v) ? (v ?? string.Empty).Trim() : string.Empty;
  }

  private static string ReadDevBypassHeader(LambdaRequest req)
  {
    var candidates = new[] { "x-dev-bypass", "X-Dev-Bypass", "x_dev_bypass", "X_DEV_BYPASS" };
    foreach (var k in candidates)
    {
      var v = PickHeader(req, k);
      if (!string.IsNullOrEmpty(v)) return v;
    }
    return string.Empty;
  }

  private static HashSet<string> ParseAllowlist(string? raw)
  {
    var s = (raw ?? string.Empty).Trim();
    if (s.Length == 0) return [];
    return s
      .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
      .Where(x => x.Length > 0)
      .ToHashSet(StringComparer.Ordinal);
  }

  private sealed record PremiumInfo(bool Ok, bool Premium, string? Env, string? ProductId, long? ExpiresAtMs, string Reason);

  private static async Task<PremiumInfo> IsPremiumFromDb(string userSub)
  {
    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null)
    {
      return new PremiumInfo(
        Ok: false,
        Premium: false,
        Env: null,
        ProductId: null,
        ExpiresAtMs: null,
        Reason: "NO_DB_POOL");
    }

    const string sql = """
      select premium_active, premium_env, product_id, expires_at_ms
      from user_premium_state
      where app_user_id = $1
      limit 1;
      """;

    var rows = await DbUtil.QueryAsync(conn, null, sql, [userSub]);
    var row = rows.Count > 0 ? rows[0] : null;

    var expiresAtMs = row is null ? null : ToLong(row.TryGetValue("expires_at_ms", out var e) ? e : null);
    var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    var premium = expiresAtMs is not null
      ? expiresAtMs > nowMs
      : row is not null && ToBool(Convert.ToString(row.TryGetValue("premium_active", out var pa) ? pa : null, CultureInfo.InvariantCulture));

    var env = row is not null ? Convert.ToString(row.TryGetValue("premium_env", out var penv) ? penv : null, CultureInfo.InvariantCulture) : null;
    var productId = row is not null ? Convert.ToString(row.TryGetValue("product_id", out var pid) ? pid : null, CultureInfo.InvariantCulture) : null;

    return new PremiumInfo(
      Ok: true,
      Premium: premium,
      Env: env,
      ProductId: productId,
      ExpiresAtMs: expiresAtMs,
      Reason: premium ? "ACTIVE" : "NOT_ACTIVE");
  }

  public static async Task<APIGatewayProxyResponse> HandlePremiumDeckUrl(
    LambdaRequest req,
    IReadOnlyDictionary<string, string> query,
    Res res,
    AuthContext auth)
  {
    if (req.Method != "GET") return res.MethodNotAllowed();

    var traceId = res.TraceId;
    var slug = SafeSlug(query.TryGetValue("slug", out var s) ? s : null);
    if (string.IsNullOrEmpty(slug)) return res.BadRequest("Missing/invalid slug");

    var apiEnv = (Environment.GetEnvironmentVariable("API_ENV") ?? string.Empty).Trim().ToLowerInvariant();
    var isProdEnv = apiEnv == "production";

    // always require login; a bearer that failed verification is a 401, not a missing login
    var badToken = Auth.RejectIfBadToken(auth, res);
    if (badToken is not null) return badToken;
    var userSub = (auth.UserSub ?? string.Empty).Trim();
    if (string.IsNullOrEmpty(userSub)) return res.Forbidden("Requires login");

    var devFlag = query.TryGetValue("dev", out var dev) && dev == "1";
    var devHeaderRaw = ReadDevBypassHeader(req);
    var devHeader = devHeaderRaw == "1";

    var allowlist = ParseAllowlist(Environment.GetEnvironmentVariable("DEV_BYPASS_ALLOWLIST_SUBS"));
    var allowlistEnabled = allowlist.Count > 0;
    var allowlisted = allowlistEnabled && allowlist.Contains(userSub);

    var allowDevBypass =
      !isProdEnv &&
      (Environment.GetEnvironmentVariable("ALLOW_DEV_PREMIUM") ?? string.Empty) == "1" &&
      (devFlag || devHeader) &&
      allowlisted;

    Log.Info(JsonSerializer.Serialize(new
    {
      traceId,
      impl = PremiumUrlImpl,
      step = "dev_bypass_eval",
      slug,
      apiEnv = string.IsNullOrEmpty(apiEnv) ? null : apiEnv,
      devFlag,
      devHeader,
      devHeaderRaw = string.IsNullOrEmpty(devHeaderRaw) ? null : devHeaderRaw,
      allowDevBypass,
      allowlistEnabled,
      allowlisted,
      ALLOW_DEV_PREMIUM = Environment.GetEnvironmentVariable("ALLOW_DEV_PREMIUM"),
      DISALLOW_SANDBOX_PREMIUM = Environment.GetEnvironmentVariable("DISALLOW_SANDBOX_PREMIUM"),
    }));

    var contentBucket = Environment.GetEnvironmentVariable("CONTENT_BUCKET");
    var premiumBucket = Environment.GetEnvironmentVariable("PREMIUM_BUCKET");

    var contentPrefix = NormalizePrefix(Environment.GetEnvironmentVariable("CONTENT_PREFIX"), "content");
    var premiumPrefix = NormalizePrefix(Environment.GetEnvironmentVariable("PREMIUM_PREFIX"), "premium");

    var expiresInSec = ToInt(Environment.GetEnvironmentVariable("PREMIUM_URL_EXPIRES_SEC"), 60);
    var s3GetTimeoutMs = ToInt(Environment.GetEnvironmentVariable("S3_GET_TIMEOUT_MS"), 6000);
    var s3BodyTimeoutMs = ToInt(Environment.GetEnvironmentVariable("S3_BODY_TIMEOUT_MS"), 6000);

    if (string.IsNullOrEmpty(contentBucket)) return res.BadRequest("CONFIG_ERROR", "Missing env CONTENT_BUCKET");
    if (string.IsNullOrEmpty(premiumBucket)) return res.BadRequest("CONFIG_ERROR", "Missing env PREMIUM_BUCKET");

    PremiumInfo? premiumInfo = null;
    if (!allowDevBypass)
    {
      try
      {
        Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "premium_check_start", userSub }));
        premiumInfo = await IsPremiumFromDb(userSub);
        Log.Info(JsonSerializer.Serialize(new
        {
          traceId,
          impl = PremiumUrlImpl,
          step = "premium_check_done",
          premium = premiumInfo.Premium,
          premiumEnv = premiumInfo.Env,
          expiresAtMs = premiumInfo.ExpiresAtMs,
        }));

        if (!premiumInfo.Premium) return res.Forbidden("Requires premium entitlement");

        if (ToBool(Environment.GetEnvironmentVariable("DISALLOW_SANDBOX_PREMIUM")))
        {
          if (string.Equals((premiumInfo.Env ?? string.Empty).Trim(), "sandbox", StringComparison.OrdinalIgnoreCase))
          {
            return res.Forbidden("Sandbox premium not allowed in this environment");
          }
        }
      }
      catch (Exception ex)
      {
        Log.Error("premium check failed:", ex);
        return res.Error500(ex);
      }
    }
    else
    {
      Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "dev_bypass_enabled", slug, userSub }));
    }

    try
    {
      // Reuse the AWS SDK client across invocations (saves connection/handler setup).
      var s3 = S3Lazy.Value;

      // 1) load manifest
      var manifestKey = $"{contentPrefix}/manifest.json";
      Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "s3_get_manifest_start", manifestKey }));

      string manifestText;
      {
        using var getCts = new CancellationTokenSource(TimeSpan.FromMilliseconds(s3GetTimeoutMs));
        using var obj = await s3.GetObjectAsync(
          new GetObjectRequest { BucketName = contentBucket, Key = manifestKey },
          getCts.Token);

        Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "s3_get_manifest_done" }));

        // 2) read body safely
        Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "manifest_read_start" }));

        using var bodyCts = new CancellationTokenSource(TimeSpan.FromMilliseconds(s3BodyTimeoutMs));
        using var sr = new StreamReader(obj.ResponseStream);
        manifestText = await sr.ReadToEndAsync(bodyCts.Token);
      }

      Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "manifest_read_done", bytes = manifestText.Length }));

      using var manifestDoc = JsonDocument.Parse(manifestText);
      var manifestRoot = manifestDoc.RootElement;

      // 3) find deck
      var deck = FindDeck(manifestRoot, slug);
      if (deck is null) return res.NotFound($"Deck not found in manifest: {slug}");

      var downloadMode = GetStringLower(deck.Value, "downloadMode");
      var tier = GetStringLower(deck.Value, "tier");

      if (!(downloadMode == "auth" || tier == "premium"))
      {
        return res.BadRequest($"Not a premium deck: {slug}");
      }

      var buildId = GetString(deck.Value, "buildId") ?? GetString(deck.Value, "version") ?? string.Empty;
      buildId = buildId.Trim();
      if (string.IsNullOrEmpty(buildId) || buildId == "coming") return res.BadRequest($"Deck not publishable: {slug}");

      // 4) presign from PREMIUM bucket
      var premiumKey = $"{premiumPrefix}/decks/{slug}/builds/{buildId}/deck.json";
      Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "presign_start", slug, buildId, premiumKey, allowDevBypass }));

      var url = s3.GetPreSignedURL(new GetPreSignedUrlRequest
      {
        BucketName = premiumBucket,
        Key = premiumKey,
        Verb = HttpVerb.GET,
        Expires = DateTime.UtcNow.AddSeconds(expiresInSec),
        ResponseHeaderOverrides = new ResponseHeaderOverrides
        {
          ContentType = "application/json",
          CacheControl = "no-cache",
        },
      });

      Log.Info(JsonSerializer.Serialize(new { traceId, impl = PremiumUrlImpl, step = "presign_done" }));

      return res.Ok(new
      {
        slug,
        buildId,
        expiresInSec,
        url,
        devBypass = allowDevBypass,
      });
    }
    catch (Exception ex)
    {
      Log.Error("handlePremiumDeckUrl error:", ex);
      return res.Error500(ex);
    }
  }

  private static long? ToLong(object? v)
  {
    if (v is null) return null;
    if (v is long l) return l;
    if (v is int i) return i;
    if (v is decimal d) return (long)d;
    if (v is double dd) return (long)Math.Floor(dd);
    var s = Convert.ToString(v, CultureInfo.InvariantCulture);
    return long.TryParse(s, NumberStyles.Integer, CultureInfo.InvariantCulture, out var n) ? n : null;
  }

  private static string NormalizePrefix(string? p, string defName)
  {
    var s = (p ?? defName).Trim();
    s = s.TrimStart('/');
    s = s.TrimEnd('/');
    return string.IsNullOrEmpty(s) ? defName : s;
  }

  private static JsonElement? FindDeck(JsonElement manifestRoot, string slugLower)
  {
    if (!manifestRoot.TryGetProperty("decks", out var decks) || decks.ValueKind != JsonValueKind.Array) return null;
    foreach (var d in decks.EnumerateArray())
    {
      if (d.ValueKind != JsonValueKind.Object) continue;
      var s = GetString(d, "slug");
      if (!string.IsNullOrEmpty(s) && string.Equals(s.Trim(), slugLower, StringComparison.OrdinalIgnoreCase))
      {
        return d;
      }
    }
    return null;
  }

  private static string? GetString(JsonElement obj, string prop)
  {
    return obj.TryGetProperty(prop, out var el) && el.ValueKind != JsonValueKind.Null && el.ValueKind != JsonValueKind.Undefined
      ? el.ToString()
      : null;
  }

  private static string GetStringLower(JsonElement obj, string prop)
  {
    return (GetString(obj, prop) ?? string.Empty).Trim().ToLowerInvariant();
  }
}
